compiler/src/dump/extract-routes.ts

```diff
@@ -0,0 +1,550 @@
+        node.children.push(oldNode)
+
+        node.name = ''
+        node.path = match
+        if (child.path.length > 0) {
+          node.children.push(child)
+        }
+      }
+    }
+
+    // if the url is lengthier than the current node we have children to populate/match
+    if (url.length > node.path.length) {
+      let found: boolean = false
+
+      // first we iterate over existing children to check if there's a match
+      for (const child of node.children) {
+        if (child.path[0] === rest[0]) {
+          found = true
+          insert(child, rest, name)
+        }
+      }
+
+      // if there's no match found we create a new child to this node
+      if (!found) {
+        const n = new Node()
+        insert(n, rest, name)
+        node.children.push(n)
+      }
+    }
+  } else if (node.path.length === 0) {
+    node.path = url
+    const match = matches(url, node.path)
+    const rest = url.slice(match.length)
+
+    node.path = match
+    if (rest.length > 0) {
+      insert(node, rest, name)
+    }
+
+    if (node.path.startsWith('{')) {
+      node.isVariable = true
+    }
+  } else if (url[0] !== node.path[0]) {
+    childrenPresent: { // eslint-disable-line
+      for (const nodeChildren of node.children) {
+        if (nodeChildren.path[0] === url[0]) {
+          insert(nodeChildren, url, name)
+          break childrenPresent // eslint-disable-line
+        }
+      }
+      const child = new Node()
+      insert(child, url, name)
+      node.children.push(child)
+    }
+  }
+
+  if (node.path === url) {
+    node.name = name
+  } else if (node.children.length === 1 && node.children[0].isVariable) {
+    node.name = name
+  } else if (node.isVariable && node.name.length === 0 && url.startsWith('{') && url.endsWith('}')) {
+    node.name = name
+  }
+}
+
+function extractRoutes (inputModel: Model): Trees {
+  const t = new Trees()
+  for (const endpoint of inputModel.endpoints) {
+    for (const url of endpoint.urls) {
+      for (const method of url.methods) {
+        t.Insert(method, url.path, endpoint.name)
+      }
+    }
+  }
+
+  return t
+}
+
+async function extractRoutesFromFiles (outPath: string): Promise<void> {
+  const v8Spec = await readFile(
+    V8,
+    { encoding: 'utf8' }
+  )
+
+  const data = await fetch(V7)
+  const v7Spec = await data.text()
+
+  const versions = new Map<number, string>()
+  versions.set(7, v7Spec)
+  versions.set(8, v8Spec)
+
+  const forest = new Forest()
+
+  versions.forEach(function (spec, version) {
+    const inputModel = JSON.parse(spec)
+    const routes = extractRoutes(inputModel)
+    forest.byVersion.set(version, routes)
+  })
+  forest.byVersion.set(0, defaultRoutes())
+
+  const str = serializeForest(forest)
+
+  await writeFile(outPath, str)
+}
+
+extractRoutesFromFiles(outputPath)
+  .catch(reason => console.error(reason))
+  .finally(() => console.log('Routes extraction complete.', outputPath))
+
+function defaultRoutes (): Trees {
```

Comment:
This defaultRoutes routine doesn't need to be kept up-to-date? Is this only used when we don't know the version of the server or for something else?

Reply:
Yep, that's the one! I've extracted that list from existing targets.